File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	// Inicio el contador de total de bytes enviados
	sentBytes := 0
	// Bucle para enviar todos los bytes
	for sentBytes < len(bytes) {
		// Envio los bytes restantes a partir del indice del ultimo byte enviado
		// n: cantidad de bytes enviados
		n, err := socket.Write(bytes[sentBytes:])
		if err != nil {
			return err
		}
		// Aumento el contador de total de bytes enviados
		sentBytes += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	receiveBuffer := make([]byte, size)
	// Inicio el contador de total de bytes recibidos
	totalBytesReceived := 0
	// Bucle para recibir todos los bytes
	for totalBytesReceived < size {
		// Recibo los bytes restantes a partir del indice del ultimo byte recibido
		// receiveBuffer[totalBytesReceived:]: cantidad de bytes restantes por recibir
		// bytesRead: cantidad de bytes recibidos
		bytesRead, err := socket.Read(receiveBuffer[totalBytesReceived:])
		if err != nil {
			return nil, err
		}
		// Aumento el contador de total de bytes recibidos
		totalBytesReceived += bytesRead
	}
	return receiveBuffer, nil
}

File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size: int):
    # Inicio contador de cantidad de bytes recibidos
    totalReceivedBytes = 0
    # Creo una variable vacio de tipo bytes para guardar los bytes recibidos
    buffer = b""
    # Bucle para recibir todos los bytes
    while totalReceivedBytes < size:
        # Recibo los bytes restantes a partir del indice del ultimo byte recibido
        # size - receivedBytes: cantidad de bytes restantes por recibir
        receivedBytes = socket.recv(size - totalReceivedBytes)
        # Si no se reciben bytes, devuelvo None
        if not receivedBytes:
            return None
        # Agrego los bytes recibidos al buffer
        buffer += receivedBytes
        # Aumento el contador de total de bytes recibidos
        totalReceivedBytes += len(receivedBytes)
          
    return buffer


def send_all(socket: socket.socket, data: bytes):
    # Inicio contador de cantidad de bytes enviados
    totalSentBytes = 0
    # Bucle para enviar todos los bytes
    while totalSentBytes < len(data):
        # Envio los bytes restantes a partir del indice del ultimo byte enviado
        # data[totalSentBytes:]: cantidad de bytes restantes por enviar
        # sentBytes: cantidad de bytes enviados
        sentBytes = socket.send(data[totalSentBytes:])
        # Si no se pudo realizar el envio, se reintenta
        if sentBytes == 0:
            continue
        # Aumento el contador de total de bytes enviados
        totalSentBytes += sentBytes

    return totalSentBytes
